### PR TITLE
Initial share provider API and UI

### PR DESCRIPTION
**Pull request by joyceerhl**
For https://github.com/microsoft/vscode/issues/176316

Introduces a share provider API and a top-level Share... command and icon in the command center toolbar. There's currently only one provider (from the builtin GitHub extension), more to come. Everything is hidden behind an unofficial setting `workbench.experimental.share.enabled`, demo:

https://github.com/microsoft/vscode/assets/30305945/6931b73b-ee28-42de-933e-ffc8595463f6

Screenshot:
image

### CODE REVIEW

File: src/vs/platform/dialogs/common/dialogs.ts [hunk 1]
```diff
@@ -154,7 +154,12 @@ export interface IPromptResultWithCancel<T> extends IPromptResult<T> {
 
 export type IDialogResult = IConfirmationResult | IInputResult | IPromptResult<unknown>;
 
-export type DialogType = 'none' | 'info' | 'error' | 'question' | 'warning';
+export type DialogType = 'none' | 'info' | 'error' | 'question' | 'warning' | 'success';
```
Comment (joyceerhl) [inline src/vs/platform/dialogs/common/dialogs.ts:157]: @bpasero this PR introduces a success/confirmation dialog to support a Share experience, given prior art in Powerpoint/Word, e.g.:
image

This also addresses a pain point where copy link actions today do not display success/confirmation and can fail silently. I would appreciate your review of the dialog-related changes here please--thank you!

**Comment by abhijit-chikane** [comment]
I think clicking on share from command palette should directly open that dialog

**Comment by joyceerhl** [comment]
@abhijit-chikane yes, it does, I just didn't include that in the recording (the recording shows that Share... exists in the palette and that there is a button to access it which does the same thing)

**Review by bpasero — COMMENTED** [review]
I find the introduction of a `Success` severity strange tbh. I think the current values for `Severity` make sense, but `Success` is not a severity for me.

File: src/vs/workbench/contrib/files/electron-sandbox/fileActions.contribution.ts [hunk 2]
```diff
@@ -71,6 +71,12 @@ MenuRegistry.appendMenuItem(MenuId.OpenEditorsContext, {
 	command: revealInOsCommand,
 	when: REVEAL_IN_OS_WHEN_CONTEXT
 });
+MenuRegistry.appendMenuItem(MenuId.OpenEditorsContextShare, {
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/electron-sandbox/fileActions.contribution.ts:74]: Fyi the "Open Editors" view is hidden by default.

Comment (joyceerhl) [reply src/vs/workbench/contrib/files/electron-sandbox/fileActions.contribution.ts:74]: :cool: I'd still like Share to appear for Open Editors if the view does happen to be visible (it is one of several places where Share should IMO appear).

**Comment by joyceerhl** [comment]
@bpasero would it help to rename `Success` to `Confirmation`, or to rename `Severity` to `Reason`?

**Comment by bpasero** [comment]
I think severity is used in many contexts, but for dialogs specifically maybe something else should be used. Here is what Electron calls it:

https://github.com/microsoft/vscode/blob/adbcbfa5d06a268f338c00ecb15e7860261e5b63/src/vs/base/parts/sandbox/common/electronTypes.ts#L19-L25

**Comment by joyceerhl** [comment]
@bpasero would it make sense for `Success` to be a subtype of the `info` type, which is internally treated differently for dialogs just to distinguish what codicon to use?

**Comment by bpasero** [comment]
@joyceerhl can you talk with SteVen as component owner of custom dialogs? Maybe someone else took that over, not sure, he would know.

File: src/vs/workbench/browser/parts/dialogs/dialogHandler.ts [hunk 3]
```diff
@@ -70,7 +70,7 @@ export class BrowserDialogHandler extends AbstractDialogHandler {
 
 		const buttons = this.getInputButtons(input);
 
-		const { button, checkboxChecked, values } = await this.doShow(input.type ?? 'question', input.message, buttons, input.detail, buttons.length - 1, input?.checkbox, input.inputs);
+		const { button, checkboxChecked, values } = await this.doShow(input.type ?? 'question', input.message, buttons, input.detail, buttons.length - 1, input?.checkbox, input.inputs, typeof input.custom !== 'boolean' ? input.custom : undefined);
```
[the remark this answers is not in the document]
Comment (joyceerhl) [reply src/vs/workbench/browser/parts/dialogs/dialogHandler.ts]: Moved this change into https://github.com/microsoft/vscode/pull/183160

File: src/vs/base/browser/ui/dialog/dialog.css [hunk 4]
```diff
@@ -60,6 +60,7 @@
 /** Dialog: Message Container */
 .monaco-dialog-box .dialog-message-row .dialog-message-container {
 	display: flex;
+	flex: 1;
```
[the remark this answers is not in the document]
Comment (joyceerhl) [reply src/vs/base/browser/ui/dialog/dialog.css]: Moved this change into https://github.com/microsoft/vscode/pull/183160

**Review by connor4312 — APPROVED** [review]

**Pull request merged**